components/brave_wallet/browser/ens_resolver_task.cc

```diff
@@ -0,0 +1,475 @@
+}
+
+}  // namespace
+
+OffchainLookupData::OffchainLookupData() = default;
+OffchainLookupData::OffchainLookupData(const OffchainLookupData&) = default;
+OffchainLookupData::OffchainLookupData(OffchainLookupData&&) = default;
+OffchainLookupData& OffchainLookupData::operator=(const OffchainLookupData&) =
+    default;
+OffchainLookupData& OffchainLookupData::operator=(OffchainLookupData&&) =
+    default;
+OffchainLookupData::~OffchainLookupData() = default;
+
+class ScopedWorkOnTask {
+ public:
+  explicit ScopedWorkOnTask(EnsGetEthAddrTask* task) : task_(task) {}
+  ~ScopedWorkOnTask() { task_->WorkOnTask(); }
+
+ private:
+  raw_ptr<EnsGetEthAddrTask> task_ = nullptr;
+};
+
+EnsGetEthAddrTask::EnsGetEthAddrTask(
+    JsonRpcServiceBase* json_rpc_service_base,
+    scoped_refptr<network::SharedURLLoaderFactory> url_loader_factory,
+    std::vector<uint8_t> ens_call,
+    const std::string& domain,
+    const std::vector<uint8_t>& dns_encoded_name,
+    const GURL& network_url)
+    : json_rpc_service_base_(json_rpc_service_base),
+      url_loader_factory_(url_loader_factory),
+      ens_call_(std::move(ens_call)),
+      domain_(domain),
+      dns_encoded_name_(dns_encoded_name),
+      network_url_(network_url) {}
+
+EnsGetEthAddrTask::~EnsGetEthAddrTask() = default;
+
+void EnsGetEthAddrTask::WorkOnTask() {
+  if (resolve_result_) {
+    json_rpc_service_base_->OnEnsResolverTaskDone(
+        this, std::move(resolve_result_.value()),
+        mojom::ProviderError::kSuccess, "");
+    // `this` is not valid here
+    return;
+  }
+  if (error_ || error_message_) {
+    json_rpc_service_base_->OnEnsResolverTaskDone(
+        this, {}, error_.value_or(mojom::ProviderError::kInternalError),
+        error_message_.value_or(""));
+    // `this` is not valid here.
+    return;
+  }
+
+  if (!resolver_) {
+    FetchEnsResolver();
+    return;
+  }
+
+  if (!supports_ensip_10_) {
+    FetchEnsip10Support();
+    return;
+  }
+
+  if (!supports_ensip_10_.value()) {
+    FetchEnsRecord();
+    return;
+  }
+
+  if (!offchain_lookup_data_) {
+    FetchWithEnsip10Resolve();
+    return;
+  }
+
+  if (offchain_lookup_attemps_left_ > 0) {
+    FetchOffchainData();
+    return;
+  }
+
+  json_rpc_service_base_->OnEnsResolverTaskDone(
+      this, {}, mojom::ProviderError::kInternalError,
+      l10n_util::GetStringUTF8(IDS_WALLET_INTERNAL_ERROR));
+  // `this` is not valid here.
+}
+
+void EnsGetEthAddrTask::FetchEnsResolver() {
+  DCHECK(!resolver_);
+  const std::string contract_address =
+      GetEnsRegistryContractAddress(brave_wallet::mojom::kMainnetChainId);
+
+  std::string call_data = ens::Resolver(domain_);
+
+  auto internal_callback =
+      base::BindOnce(&EnsGetEthAddrTask::OnFetchEnsResolverDone,
+                     weak_ptr_factory_.GetWeakPtr());
+  json_rpc_service_base_->RequestInternal(
+      eth::eth_call(contract_address, call_data), true, network_url_,
+      std::move(internal_callback), base::NullCallback());
```

Comment:
is something passing the JSON from the network through the JSON sanitizer before `internal_callback` gets called?

Reply:
Good catch, it's missing here. Will fix.
Thanks!

Reply:
Actually `json_rpc_service_base_->RequestInternal` will sanitize json internally, so nothing to add here.